src/NuGet.Clients/NuGet.Console/ScriptExecutor.cs

```diff
@@ -178,7 +187,11 @@ private async Task ExecuteScriptCoreAsync(ScriptExecutionRequest request)
             // to switch to powershell pipeline execution thread. In order not to block the UI thread,
             // go off the UI thread. This is important, since, switches to UI thread,
             // using SwitchToMainThreadAsync will deadlock otherwise
-            await Task.Run(() => host.Execute(console, request.BuildCommand(), request.BuildInput()));
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+            _pumpingJTF.Run(async () =>
```

Comment:
I'm not sure this is the best solution we have. 

Why do we need to switch to the UI thread explicitly? 
Wouldn't running it with the pumping JTF be enough? 

Do we need to block on the UI thread?

Reply:
yes, we have to run the PumpingJTF on the UI thread, that's the only way it can work. Running the PumpingJTF on a background thread, while something else blocks the UI thread with a normal JTF blocks the .NET runtime from doing COM RPC, which needs the UI thread.

Reply:
What PumpingJTF does different form the regular JTF is that it turns off the message pump synchronization. 
So a PumpingJTF essentially makes the UI thread version, work the same way as the non-UI thread one.

Anyway, what if we run async and *not* block the UI thread? 

That way we're not introducing the UI delay in the PMC scenarios (given that Roslyn already blocks the UI thread in the extensibility scenarios, it won't make anything worse.

Reply:
Using `await _pumpingJTF.RunAsync(...)` instead of `_pumpingJTF.Run(...)` means that the pumping JTF disappears from the main thread's call stack. Therefore when Powershell tries to access a DTE object, it's unable to switch to the UI thread. The deadlock remains.

I agree it's not ideal for us to use pumping JTF on the UI thread when we install via PMC or PM UI, because we don't block the UI thread in our own code. I investigated passing a bool from `VsPackageInstaller.InstallPackage` all the way up the call stack to here, `ScriptExecutor`, so we only use pumping JTF when it originates from a VS extensibility API call where we don't control whether or not the caller is blocking the UI thread.  However, it had massive impact, including on the public APIs in NuGet.PackageManagement.

Another option is to revert this PR's changes and simply switch to the UI thread before running PumpingJTF.Run in VsPackageInstaller.InstallPackage, but that means that the entire package install operation will be blocking the UI thread, not just the powershell script. Considering not all packages have an install.ps1, and of those that do, it only runs when the target project uses pacakges.config, it's skipped when the project uses PackageReference, this would mean that using PumpingJTF for all VS extensibility API calls will have a UI thread impact for a significant number of API calls when it's never needed.

Reply:
Discused offline.